sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts

```diff
@@ -0,0 +1,157 @@
+async function sessionAccepted(sessionId: string) {
+  console.log(`[${sessionId}] will start processing...`);
+}
+
+// Called by the SessionReceiver when a message is received.
+// This is passed as part of the handlers when calling `SessionReceiver.subscribe()`.
+async function processMessage(msg: ReceivedMessageInfo) {
+  console.log(`[${msg.sessionId}] received message with body ${msg.body}`);
+}
+
+// Called by the SessionReceiver when an error occurs.
+// This will be called in the handlers we pass in `SessionReceiver.subscribe()`
+// and by the sample when we encounter an error opening a session.
+function processError(err: Error, sessionId?: string) {
+  if (sessionId) {
+    console.log(`Error when receiving messages from the session ${sessionId}: `, err);
+  } else {
+    console.log(`Error when creating the receiver for next available session`, err);
+  }
+}
+
+// Called if we are closing a session.
+// `reason` will be:
+// * 'error' if we are closing because of an error(the error will be delivered
+//   to `processError` above)
+// * 'idle_timeout' if `sessionIdleTimeoutMs` milliseconds pass without
+//   any messages being received (ie, session can be considered empty).
+// NOTE: This function is used only in the sample and is not part of the Service Bus library.
+async function sessionClosed(reason: "error" | "idle_timeout", sessionId: string) {
+  console.log(`[${sessionId}] was closed because of ${reason}`);
+}
+
+// utility function to create a timer that can be refreshed
+function createRefreshableTimer(timeoutMs: number, resolve: Function): () => void {
+  let timer: any;
+
+  return () => {
+    clearTimeout(timer);
+    timer = setTimeout(() => resolve(), timeoutMs);
+  };
+}
+
+// Queries Service Bus for the next available session and processes it.
+async function receiveFromNextSession(queueClient: QueueClient): Promise<void> {
+  let sessionReceiver = queueClient.createReceiver(ReceiveMode.peekLock, {
+    sessionId: undefined
+  });
+
+  try {
+    await sessionReceiver.getState();
+  } catch (err) {
+    if (
+      (err as MessagingError).condition === "com.microsoft:session-cannot-be-locked" ||
+      (err as MessagingError).condition === "amqp:operation-timeout"
+    ) {
+      console.log(`INFO: no available sessions, sleeping for ${delayOnErrorMs}`);
+    } else {
+      processError(err, undefined);
+    }
+
+    await delay(delayOnErrorMs);
+    return;
+  }
+
+  await sessionAccepted(sessionReceiver.sessionId);
+
+  const sessionFullyRead = new Promise((resolveSessionAsFullyRead, rejectSessionWithError) => {
+    const refreshTimer = createRefreshableTimer(
+      sessionIdleTimeoutSeconds * 1000,
+      resolveSessionAsFullyRead
+    );
+    refreshTimer();
+
+    abortController.signal.addEventListener("abort", function abortProcessing() {
+      rejectSessionWithError(new Error("Process terminated by user."));
+      abortController.signal.removeEventListener("abort", abortProcessing);
```

Comment:
We should also remove the event listener when `sessionFullyRead` is resolved or rejected, else we will keep adding listeners for each session, but remove them only if the abort signal was fired

Reply:
I've added in removals for each point where it could exit now.